商品出品機能

shun7603 opened the pull request:
# What
商品出品機能の実装

#Why
商品出品機能を設けることで、ユーザーは商品を出品し、他のユーザーに販売できるようにする。これにより、ユーザー同士の売買が可能になる。

ログイン状態は商品出品ページへ遷移できる↓
https://gyazo.com/7edfe8159276e4f25420116705d5cc97

価格が入力されると販売手数料と販売利益が表示される↓
https://gyazo.com/28408b16e0b4b7f27ea1d8d8d8b247ab

必要な情報を適切に入力して「出品する」を押すと、商品情報がデータベースに保存される↓
https://gyazo.com/bcf6c6ceffed11c51dbf17a7100e195e

入力に問題がある状態で「出品する」を押すと、情報は保存されず、出品ページに戻りエラーメッセージが表示される↓
https://gyazo.com/54ccb66a75a67053b2c42f7a5baea3df

ログアウト状態の場合は、商品出品ページへ遷移しようとすると、ログインページへ遷移する動画
https://gyazo.com/5232207eb65ea2f8e87786a77b1b1173

テスト結果の画像↓
https://gyazo.com/e1dde7036eecbd37fc98b8de659e0583

以下、よろしくお願いいたします！

shun7603 committed:
商品出品機能

itemモデル作成
activeHash設定
コントローラー設定
ビューのヘルパーメソッド設定
ルーティング設定

shun7603 committed:
商品出品機能

itemモデル設定
マイグレーション設定
itemコントローラー設定（new,create)
itemルーティング設定
priceの自動計算する設定（JS)
エラーハンドリング表示設定

wad-sho reviewed (commented):
実装お疲れ様です。レビュー依頼確認しました。

拝見したところ、商品一覧表示機能の内容が実装されているようです。
app/views/items/index.html.erbとコントローラーのindexアクションにおける、商品一覧表示機能の内容は元に戻しましょう。

元に戻した上で、再度レビュー依頼お願いいたします。

shun7603 committed:
商品出品機能

商品詳細機能の内容をもとに戻した

shun7603 committed:
商品機能実装

商品詳細機能 ビューを元に戻しました。

ushi10 commented on app/javascript/item_price.js:
```diff
@@ -0,0 +1,25 @@
+document.addEventListener('turbo:load', () => {
```
商品の出品に失敗して出品ページに戻った際もjavascriptが動作するよう、renderメソッドに対応したイベントを追加しましょう。
以下カリキュラムのSTEP2、手順6をご参考ください。

実装に悩んだときのヒント
商品出品機能
https://master.tech-camp.in/v2/curriculums/8138

ushi10 commented on app/models/item.rb:
```diff
@@ -0,0 +1,57 @@
+  belongs_to :shipping_fee
+  belongs_to :shipping_day
+
+  has_one :order
```
```ruby
has_one :order
```
上記は購入機能時に実装するため現時点ではコメントアウトをしておきましょう。

ushi10 commented on app/models/item.rb:
```diff
@@ -0,0 +1,57 @@
+      return
+    end
+
+    unless price_before_type_cast.to_s.match?(/\A[0-9]+\z/)
```
integer型に正規表現は適用されないため、バリデーションを用いて「整数値のみ受け付ける」という制限をかけましょう。

ushi10 commented on app/controllers/items_controller.rb:
```diff
@@ -1,4 +1,23 @@
 class ItemsController < ApplicationController
-  def index
```
indexアクションそのものは残しておき、内部の処理のみ削除しておきましょう。
アクションがない場合、トップページが表示されなくなる場合があるためです。

ushi10 commented on app/views/items/new.html.erb:
```diff
@@ -5,10 +5,10 @@
     <%# インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
-    <%# render 'shared/error_messages', model: f.object %>
+    <%= render 'shared/error_messages', model: f.object %>
     <%# //インスタンスを渡して、エラー発生時にメッセージが表示されるようにしましょう。%>
```
10,12目の実装案内文は不要なため削除しましょう。
もし必要であれば第三者でもわかるように「残す目的」「なぜ必要か」という理由などを、コメントとして記載しておきましょう。

ushi10 reviewed (commented):
レビュー依頼確認しました。
コメントいたしましたのでご確認ください。

修正、Rubocop実行後、コミットプッシュを確認いただき再度レビュー依頼のご提出をお願いいたします。

shun7603 committed:
商品出品機能実装

出品エラー発生後でもJSが動作するようにイベントを追加
orderのアソシエーションをコメントアウト
priceのバリデーション再設定
indexアクション追加
不要なコメント削除

mizuki-ueda reviewed (commented):
レビュー依頼確認しました！

LGTMとさせていただきます！
実装、お疲れ様でした！

マージできましたら、ローカルのmainブランチへ、忘れずにプルしましょう。
そのあと、新しいブランチを作成し、次の実装に取り組んでください。

引き続き、実装頑張っていきましょう。
応援しております！
